recipe-server/client/control/components/MultiPicker.js

```diff
@@ -51,88 +44,65 @@ export default class MultiPicker extends React.Component {
       value,
     } = this.props;
 
-    const selectedOptions = options.filter(option =>
-      value.indexOf(option.value) === -1);
-
-    let displayedOptions = [].concat(selectedOptions)
-      .map(option => option.value)
-      .map(val =>
-        (!val ? null : { ...this.convertValueToObj(val) }))
-      .filter(x => x);
-
     const {
       filterText,
     } = this.state;
 
-    if (filterText) {
-      displayedOptions = displayedOptions.filter(option =>
-        JSON.stringify(option).indexOf(filterText) > -1);
-    }
+    let selectedOptions = options.filter(option => value.indexOf(option.value) === -1);
 
-    return displayedOptions;
-  }
-
-  handleApplyOption(event) {
-    event.persist();
-
-    if (!this.availableRef) {
-      return;
-    }
-
-    const {
-      value,
-      onChange,
-    } = this.props;
-
-    let selectedFilters = MultiPicker.getActiveSelectOptions(this.availableRef);
+    if (filterText) {
+      const searchText = filterText.toLowerCase();
 
-    if (!selectedFilters || selectedFilters.length === 0) {
-      selectedFilters = this.getDisplayedOptions();
-      selectedFilters = selectedFilters.map(option => option.value);
+      selectedOptions = selectedOptions.filter(option => {
+        const searchValues = JSON.stringify(Object.values(option)).toLowerCase();
+        return searchValues.indexOf(searchText) > -1;
+      });
     }
 
-    const newOptions = []
-      .concat(value || [])
-      .concat(selectedFilters);
-
-    // clear user input
-    this.availableRef.value = null;
-
-    onChange(newOptions.join(','));
+    return selectedOptions;
   }
 
-  handleRemoveOption(event) {
-    event.persist();
-    if (!this.selectedRef) {
-      return;
-    }
-
-    const {
-      value,
-      onChange,
-    } = this.props;
+  handleTextChange(event) {
+    const { value } = event.target;
 
-    const selectedFilters = MultiPicker.getActiveSelectOptions(this.selectedRef);
-    let newOptions = []
-      .concat(value || [])
-      .filter(val => selectedFilters.indexOf(val) === -1);
+    this.setState({
+      filterText: value,
+    });
+  }
 
-    if (!selectedFilters || selectedFilters.length === 0) {
-      newOptions = [];
-    }
+  handleChangeOption(type) {
+    return () => {
+      const ref = type === 'apply' ? this.availableRef : this.selectedRef;
+      if (!type || !ref) {
+        return;
+      }
 
-    // clear the user selection
-    this.selectedRef.value = null;
+      const {
+        value = [],
+        onChange = () => {},
+      } = this.props;
+
+      // Get a list of all selected values.
+      let selection = [].concat(value);
+      // Pick up list of selected options from the dom ref
+      const refSelected = MultiPicker.getActiveSelectOptions(ref);
+
+      // Apply = take from the 'available' section and put in 'selected'
+      if (type === 'apply') {
```

Comment:
Sure! Also, I was mistaken: upon further thought, making a standalone component for the option lists makes the subclassing idea unnecessary. But I'll describe it anyway.

The basic pattern is, wherever the behavior needs to be customized, instead of an `if` statement, you call another method on the class. Then, you create a subclass for each branch of the `if` that implements that method with the customized behavior. 

Directly applying that to this example (and simplifying the classes a bit):

```js
class Base {
  handleChangeOption() {
    const ref = this.getRef();
    if (!ref) {
      return;
    }

    const {
      value = [],
      onChange = () => {},
    } = this.props;

    // Get a list of all selected values.
    let selection = [].concat(value);
    // Pick up list of selected options from the dom ref
    const refSelected = MultiPicker.getActiveSelectOptions(ref);

    selection = this.modifySelection(selection, refSelected);

    // Clear user input.
    ref.value = null;

    // Report the updated selected filters to the parent.
    onChange(selection.join(','));
  };
}

class Apply extends Base {
  getRef() {
    return this.availableRef;
  }

  modifySelection(selection, refSelected) {
    return selection.concat(refSelected)
  }
}

class Remove extends Base {
  getRef() {
    return this.selectedRef;
  }

  modifySelection(selection, refSelected) {
    return selection.filter(val => refSelected.indexOf(val) === -1);
  }
}
```

That's not the whole story though, because you need to figure out where to actually use these two subclasses to get the customized behavior. With React, it's almost always by adding a new component. 

One option would be to make the add and remove buttons components that share a base class. You'd have to pass the `value` and `onChange` props down to them, but that's fine.  Because the buttons themselves don't contain the lists, meaning you can't get a ref to them, you'd have to pass that down as a ref as well. 

---

Stepping back, the core problem here is finding the right abstraction for what we want to do. `handleChangeOptions` does two things:

1. Fetches the list of options in one of the two selection boxes.
2. Modifies the form value of selected values by either removing or adding the selected options.

Actually, these are the two things that differ between adding and removing. The rest of the shared code is mostly plumbing. 

The first item, fetching the selected options, is actually the same for both lists, just with a different ref. That's one reason why [I suggested](https://github.com/mozilla/normandy/pull/595#discussion_r106298641) that we use a single component for both lists. The component can care all it wants about refs, and it just takes a callback that reports the selected items.

If the first item is handled, the second one is easier, because the only shared logic left is calling `onChange`, and one line of duplicated code is easy to read and understand.

---

That's a bit wordy. I've never really pair programmed before, but if my spiel above is still unclear, we can work together on writing the generic filter component so I can show you what I mean. :D